lib/mongoid/traversable.rb

```diff
@@ -54,18 +53,38 @@ class << self
 
       def discriminator_value=(value)
         value ||= self.name
+        add_discriminator_mapping(value)
         super
       end
     end
 
     included do
       class_attribute :discriminator_key, instance_accessor: false
       class_attribute :discriminator_value, instance_accessor: false
-      self.discriminator_value = self.name
-
+            
       class << self
         delegate :discriminator_key, to: ::Mongoid
         prepend DiscriminatorAssignment
+        
+        # @api private
+        attr_accessor :discriminator_mapping
+      end
+
+      # Add a discriminator mapping to the parent class. This mapping is used when
+      # receiving a document to identify its class.
+      #
+      # @api private
+      def self.add_discriminator_mapping(value, klass=self)
+        self.discriminator_mapping ||= {}
+        self.discriminator_mapping[value] = klass
+        superclass.add_discriminator_mapping(value, klass) if hereditary?
+      end
+      
+      # Get the discriminator mapping from the parent class
+      #
+      # @api private
+      def self.get_discriminator_mapping(value)
+        self.discriminator_mapping[value]
```

Comment:
Is this method expected to do more in the future?

Reply:
nope, this is it, I believe... I did in the past so that's why I had it there... But I honestly kind of like it instead of dealing with the discriminator_mapping hash directly

Reply:
Scratch that, I did have to add a check that discriminator_mapping exists (i.e. when it's not hereditary)